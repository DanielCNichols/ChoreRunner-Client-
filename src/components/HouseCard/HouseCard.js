import React, { useState } from 'react';
import s from './HouseCard.module.css';
import ApiService from '../../services/api-service';
import Modal from '../Modal/Modal';
import AddMembers from '../AddMembers/AddMemberForm';
import { Link } from 'react-router-dom';
import Member from '../Member/Member';

import {
  MdSave,
  MdEdit,
  MdDelete,
  MdPersonAdd,
  MdCancel,
  MdChevronRight,
} from 'react-icons/md';

export default function HouseCard({
  house,
  handleDelete,
  handleEdit,
  handleAddMembers,
  handleEditMember,
  deleteMember,
}) {
  const [name, setName] = useState(house.name);
  const [addMember, setAddMember] = useState(false);
  const [editingMember, setEditingMember] = useState(false);
  const [editing, setEditing] = useState(false);

  const [error, setError] = useState(null);

  const handleNameChange = e => {
    e.preventDefault();

    setName(e.target.value);
  };

  const handleSubmit = async () => {
    try {
      if (!name.length) {
        throw new Error('Name is required');
      }

      let res = await ApiService.editHouseholdName(house.id, {
        name,
        id: house.id,
      });

      handleEdit(res);
      setEditing(false);
    } catch (error) {
      setError(error);
    }
  };

  const handleDeleteHousehold = async id => {
    try {
      await ApiService.deleteHousehold(id);
      handleDelete(id);
    } catch (error) {
      setError(error);
    }
  };

  const toggleAddMember = () => {
    setAddMember(!addMember);
  };

  const handleDeleteMember = async (memberId, householdId) => {
    try {
      await ApiService.deleteMember(memberId);

      deleteMember(memberId, householdId);
    } catch (error) {
      setError(error);
    }
  };

  return (
    <div className={s.houseCard}>
      {addMember && (
        <Modal titleText="Add Member Form">
          <AddMembers
            id={house.id}
            handleAddMembers={handleAddMembers}
            toggleAddMember={toggleAddMember}
          />
        </Modal>
      )}
      <>
        <div className={s.cardHeader}>
          <header>
            {editing ? (
              <input value={name} onChange={ev => handleNameChange(ev)} />
            ) : (
              <h4 className="videoGameTitles">{house.name}</h4>
            )}
          </header>
          <div className={s.buttonContainer}>
            {editing ? (
              <div className={s.editButtons}>
                <button onClick={() => setEditing(false)}>
                  <MdCancel />
                </button>
                <button onClick={() => handleSubmit()}>
                  <MdSave />
                </button>
              </div>
            ) : (
              <>
                <button onClick={() => setAddMember(true)}>
                  <MdPersonAdd />
                </button>
                <button onClick={() => setEditing(true)}>
                  <MdEdit />
                </button>
                <button onClick={() => handleDeleteHousehold(house.id)}>
                  <MdDelete />
                </button>
              </>
            )}
          </div>
        </div>

        {house.members.length ? (
          <div className={s.cardBody}>
            <p>Name</p>
            <p>Level</p>
            <p>Exp</p>

            <ul className={s.membersList}>
              {house.members.map(member => {
                return (
                  <Member
                    key={member.id}
                    showEdit={editing}
                    member={member}
                    deleteMember={handleDeleteMember}
                    editMember={handleEditMember}
                  />
                );
              })}
            </ul>
          </div>
        ) : (
          <p>You have not added any members to this household</p>
        )}
        {!!house.members.length && (
          <div className={s.detailsLink}>
            <Link to={`/household/${house.id}`}>
              Manage Tasks
              <MdChevronRight />
            </Link>
          </div>
        )}
      </>
    </div>
  );
}
