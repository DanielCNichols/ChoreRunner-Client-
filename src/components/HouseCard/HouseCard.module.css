@import '../../vars.css';

.houseCard {
  width: 100%;
  display: flex;
  flex-direction: column;
  background-color: var(--color-menus);
  color: white;
  border: 2px solid white;
  border-radius: 6px;
  margin: 20px auto;
  font-family: var(--font-body);
}

.cardHeader {
  width: 100%;
  border-bottom: 2px solid white;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.cardHeader header {
  padding: 5px 10px;
}

.buttonContainer {
  width: 50%;
  max-width: 145px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 5px 10px;
  font-size: 1.3em;
}

.buttonContainer svg {
  color: white;
}

.cardHeader h4 {
  margin: 0;
}

.cardBody {
  display: grid;
  grid-template-rows: 2;
  grid-template-columns: 1fr 30% 1fr;
  padding: 10px 20px;
}

.cardBody > p {
  grid-row: 1;
  margin: 0;
  text-align: left;
}

.cardBody p:nth-child(2) {
  text-align: center;
}

.cardBody p:nth-child(3) {
  text-align: center;
}

.membersList {
  margin: 10px 0;
  padding: 0;
  grid-column: span 3;
}

.memberDetails {
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 30% 1fr;
  text-align: left;
  margin-bottom: 10px;
}

.memberDetails p,
.progressBarContainer {
  margin: 0;
  display: flex;
  align-items: center;
}

@media screen and (min-width: 450px) {
  .houseCard {
    width: 50%;
    max-width: 400px;
  }
}

@media screen and (min-width: 800px) {
  .houseCard {
    width: 100%;
  }
}
