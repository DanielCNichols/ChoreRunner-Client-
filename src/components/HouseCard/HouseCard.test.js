import React from 'react';
import ReactDOM from 'react-dom';
import HouseCard from './HouseCard';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('HouseCard component testing', () => {
  const house = {
    name: 'test',
    id: 1,
    members: [],
  };
  const handleDelete = () => {};
  const handleEdit = () => {};
  const handleAddMembers = () => {};
  const handleEditMember = () => {};
  const deleteMember = () => {};

  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(
      <MemoryRouter>
        <HouseCard
          house={house}
          handleDelete={handleDelete}
          handleEdit={handleEdit}
          handleAddMembers={handleAddMembers}
          handleEditMember={handleEditMember}
          deleteMember={handleDelete}
        />
      </MemoryRouter>,
      div
    );

    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <HouseCard
            house={house}
            handleDelete={handleDelete}
            handleEdit={handleEdit}
            handleAddMembers={handleAddMembers}
            handleEditMember={handleEditMember}
            deleteMember={handleDelete}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
