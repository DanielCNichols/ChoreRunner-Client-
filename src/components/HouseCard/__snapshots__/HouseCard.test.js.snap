// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HouseCard component testing renders UI as expected 1`] = `
<div
  className="houseCard"
>
  <div
    className="cardHeader"
  >
    <header>
      <h4
        className="videoGameTitles"
      >
        test
      </h4>
    </header>
    <div
      className="buttonContainer"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm-9-2V7H4v3H1v2h3v3h2v-3h3v-2H6zm9 4c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </button>
      <button
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </button>
      <button
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
  <p>
    You have not added any members to this household
  </p>
</div>
`;
